js/src/common/extend.ts

```diff
@@ -1,3 +1,10 @@
+export type KeyOfType<T extends object, Type> = Exclude<
+  {
+    [K in keyof T]: T[K] extends Type ? K : never;
+  }[keyof T],
+  undefined
+>;
```

Comment:
@askvortsov1 I think this is clearer.

```suggestion
type KeyOfType<Type extends object, Match> = { [Key in keyof Type]-?: Type[Key] extends Match ? Key : never }[keyof Type];
```

1. `= { ... }` -  Create a new object type...
2. `[K in keyof Type]` - ...with all keys of `Type`
3. `-?:` -  ...and make them all required (convert `string | undefined` to `string`)
4. `Type[Key] extends Match ?` ...and set each key's type conditionally based on whether the type of the key is a subtype of the `Match` type
5. `Key : never` - ...so that the key's type becomes the key itself if its one we care about (`Key`), or it gets dropped if we don't care about it (`never`)
6. `[keyof Type]` - Then we convert the map of key to key into an array of the keys

Reply:
Dropping the `Exclude` feels cleaner, but I still think we should break this up into steps so it's more parseable.

Reply:
How, though? The only way we can split it is into KeysOfType and KeyOfType. Is that fine?